framework/src/main/java/org/checkerframework/framework/ajava/InsertAjavaAnnotations.java

```diff
@@ -0,0 +1,580 @@
+                int lastSize = cumulativeLineSizes.get(i - 1);
+                int lastLineLength = lines[i - 1].length() + System.lineSeparator().length();
+                cumulativeLineSizes.add(lastSize + lastLineLength);
+            }
+        }
+
+        @Override
+        public void defaultAction(Node node1, Node node2) {
+            if (!(node1 instanceof NodeWithAnnotations<?>)) {
+                return;
+            }
+
+            NodeWithAnnotations<?> node1WithAnnos = (NodeWithAnnotations<?>) node1;
+            if (node1 instanceof MethodDeclaration) {
+                addAnnotationOnOwnLine(node2.getBegin().get(), node1WithAnnos.getAnnotations());
+                return;
+            } else if (node1 instanceof FieldDeclaration) {
+                addAnnotationOnOwnLine(node2.getBegin().get(), node1WithAnnos.getAnnotations());
+                return;
+            }
+
+            Position position;
+            if (node2 instanceof ClassOrInterfaceType) {
+                // In a multi-part name like my.package.MyClass, annotations go directly in front of
+                // MyClass instead of the full name.
+                position = ((ClassOrInterfaceType) node2).getName().getBegin().get();
+            } else {
+                position = node2.getBegin().get();
+            }
+
+            addAnnotations(position, node1WithAnnos.getAnnotations(), 0, false);
+        }
+
+        @Override
+        public void visit(ArrayType node1, Node other) {
+            ArrayType node2 = (ArrayType) other;
+            // The second component of this pair contains a list of ArrayBracketPairs from left to
+            // right. For example, if node1 contains String[][], then the list will contain the
+            // types String[] and String[][]. To insert array annotations in the correct location,
+            // we insert them directly to the right of the end of the previous element.
+            Pair<Type, List<ArrayType.ArrayBracketPair>> node1ArrayTypes =
+                    ArrayType.unwrapArrayTypes(node1);
+            Pair<Type, List<ArrayType.ArrayBracketPair>> node2ArrayTypes =
+                    ArrayType.unwrapArrayTypes(node2);
+            // The first annotations go directly after the element type.
+            Position firstPosition = node2ArrayTypes.a.getEnd().get();
+            addAnnotations(firstPosition, node1ArrayTypes.b.get(0).getAnnotations(), 1, false);
+            for (int i = 1; i < node1ArrayTypes.b.size(); i++) {
+                Position position =
+                        node2ArrayTypes.b.get(i - 1).getTokenRange().get().toRange().get().end;
+                addAnnotations(position, node1ArrayTypes.b.get(i).getAnnotations(), 1, true);
+            }
+
+            // Visit the component type.
+            node1ArrayTypes.a.accept(this, node2ArrayTypes.a);
+        }
+
+        @Override
+        public void visit(final CompilationUnit node1, final Node other) {
```

Comment:
Minor point:  I find `final` to be clutter on these formal parameters.  It is not present on other formal parameters in the file, nor on local variables in this method.